exporter/awss3exporter/s3_writer.go

```diff
@@ -0,0 +1,87 @@
+// Copyright 2022 OpenTelemetry Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package awss3exporter // import "github.com/open-telemetry/opentelemetry-collector-contrib/exporter/awss3exporter"
+
+import (
+	"bytes"
+	"context"
+	"fmt"
+	"math/rand"
+	"strconv"
+	"time"
+
+	"github.com/aws/aws-sdk-go/aws"
+	"github.com/aws/aws-sdk-go/aws/session"
+	"github.com/aws/aws-sdk-go/service/s3/s3manager"
+)
+
+type S3Writer struct {
+}
+
+// generate the s3 time key based on partition configuration
+func getTimeKey(time time.Time, partition string) string {
+	var timeKey string
+	year, month, day := time.Date()
+	hour, minute, _ := time.Clock()
+
+	rand.Int()
+
+	if partition == "hour" {
+		timeKey = fmt.Sprintf("year=%d/month=%02d/day=%02d/hour=%02d", year, month, day, hour)
+	} else {
+		timeKey = fmt.Sprintf("year=%d/month=%02d/day=%02d/hour=%02d/minute=%02d", year, month, day, hour, minute)
+	}
+	return timeKey
+}
+
+func getS3Key(time time.Time, bucket string, keyPrefix string, partition string, filePrefix string, metadata string, fileformat string) string {
+	timeKey := getTimeKey(time, partition)
+	randomID := rand.Int()
```

Comment:
Not sure if that matters much, but perhaps it would be nice to have padded number (so always the same number of characters). Also, some encoding (even base64) could be used to shorten the random part. S3 has max object length of 1024 so we should be well within the limit, but, just in case, might be worth considering